File: typegraph/python/typegraph/graph/tg_deploy.py
# Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
# SPDX-License-Identifier: MPL-2.0

import json
import os
from dataclasses import dataclass
from typing import Dict, Optional, Union
from urllib import request

from typegraph.gen.exports.utils import QueryDeployParams
from typegraph.gen.types import Err
from typegraph.graph.shared_types import BasicAuth
from typegraph.graph.typegraph import TypegraphOutput
from typegraph.wit import ArtifactResolutionConfig, store, wit_utils


@dataclass
class TypegraphDeployParams:
    base_url: str
    artifacts_config: ArtifactResolutionConfig
    auth: Optional[BasicAuth] = None
    secrets: Optional[Dict[str, str]] = None


@dataclass
class TypegraphRemoveParams:
    base_url: str
    auth: Optional[BasicAuth] = None


@dataclass
class DeployResult:
    serialized: str
    typegate: Union[Dict[str, any], str]


@dataclass
class RemoveResult:
    typegate: Union[Dict[str, any], str]


@dataclass
class UploadArtifactMeta:
    name: str
    artifact_hash: str
    artifact_size_in_bytes: int


def tg_deploy(tg: TypegraphOutput, params: TypegraphDeployParams) -> DeployResult:
    sep = "/" if not params.base_url.endswith("/") else ""
    url = params.base_url + sep + "typegate"

    headers = {"Content-Type": "application/json"}
    if params.auth is not None:
        headers["Authorization"] = params.auth.as_header_value()
    serialized = tg.serialize(params.artifacts_config)
    tg_json = serialized.tgJson
    ref_artifacts = serialized.ref_artifacts

    # upload the referred artifacts
    # TODO: fetch all the upload urls in one request
    get_upload_url = params.base_url + sep + tg.name + "/get-upload-url"
    for artifact_hash, artifact_path in ref_artifacts:
        with open(artifact_path, "rb") as artifact:
            artifact_content = artifact.read()
            artifact = UploadArtifactMeta(
                name=os.path.basename(artifact_path),
                artifact_hash=artifact_hash,
                artifact_size_in_bytes=len(artifact_content),
            )

            artifact_json = json.dumps(artifact.__dict__).encode()
            req = request.Request(
                url=get_upload_url, method="PUT", headers=headers, data=artifact_json
            )

            response = handle_response(exec_request(req).read().decode())
            artifact_upload_url = response["uploadUrl"]

            upload_headers = {"Content-Type": "application/octet-stream"}
            if params.auth is not None:
                upload_headers["Authorization"] = params.auth.as_header_value()
            upload_req = request.Request(
                url=artifact_upload_url,
                method="PUT",
                data=artifact_content,
                headers=upload_headers,
            )
            response = request.urlopen(upload_req)
            if response.status != 200:
                raise Exception(
                    f"Failed to upload artifact {artifact_path} to typegate: {response.read()}"
                )

    # deploy the typegraph
    res = wit_utils.gql_deploy_query(
        store,
        params=QueryDeployParams(
            tg=tg_json,
            secrets=[(k, v) for k, v in (params.secrets or {}).items()],
        ),
    )

    if isinstance(res, Err):
        raise Exception(res.value)

    req = request.Request(
        url=url,
        method="POST",
        headers=headers,
        data=res.value.encode(),
    )

    return DeployResult(
        serialized=tg_json,
        typegate=handle_response(exec_request(req).read().decode()),
    )


def tg_remove(tg: TypegraphOutput, params: TypegraphRemoveParams):
    sep = "/" if not params.base_url.endswith("/") else ""
    url = params.base_url + sep + "typegate"

    headers = {"Content-Type": "application/json"}
    if params.auth is not None:
        headers["Authorization"] = params.auth.as_header_value()

    res = wit_utils.gql_remove_query(store, [tg.name])

    if isinstance(res, Err):
        raise Exception(res.value)

    req = request.Request(
        url=url,
        method="POST",
        headers=headers,
        data=res.value.encode(),
    )

    return RemoveResult(typegate=handle_response(exec_request(req).read().decode()))


# simple wrapper for a more descriptive error
def exec_request(req: any):
    try:
        return request.urlopen(req)
    except request.HTTPError as res:
        # Note: 400 status and such, the response body
        # is hidden within the exception and can be consumed through .read()
        return res
    except Exception as e:
        raise Exception(f"{e}: {req.full_url}")


def handle_response(res: any):
    try:
        return json.loads(res)
    except Exception as _:
        return res

File: src/typegraph/deno/src/tg_deploy.ts
// Copyright Metatype OÜ, licensed under the Mozilla Public License Version 2.0.
// SPDX-License-Identifier: MPL-2.0

import type { MigrationAction, SerializeParams } from "./gen/core.ts";
import { ArtifactUploader } from "./tg_artifact_upload.ts";
import type { TypegraphOutput } from "./typegraph.ts";
import { sdkUtils } from "./sdk.ts";
import { execRequest } from "./utils/func_utils.ts";
import { log } from "./io.ts";

export class BasicAuth {
  constructor(
    public username: string,
    public password: string,
  ) {}
  asHeaderValue(): string {
    return `Basic ${btoa(this.username + ":" + this.password)}`;
  }
}

export interface TypegateConnectionOptions {
  url: string;
  auth?: BasicAuth;
}

export interface TypegraphDeployParams {
  typegate: TypegateConnectionOptions;
  typegraphPath: string;
  prefix?: string;
  secrets?: Record<string, string>;
  migrationsDir?: string;
  migrationActions?: Record<string, MigrationAction>;
  defaultMigrationAction?: MigrationAction;
}

export interface TypegraphRemoveParams {
  typegate: TypegateConnectionOptions;
}

export interface DeployResult {
  serialized: string;
  // deno-lint-ignore no-explicit-any
  response: Record<string, any>;
}

export interface RemoveResult {
  // deno-lint-ignore no-explicit-any
  typegate: Record<string, any> | string;
}

export interface ArtifactMeta {
  typegraphName: string;
  relativePath: string;
  hash: string;
  sizeInBytes: number;
}

export async function tgDeploy(
  typegraph: TypegraphOutput,
  params: TypegraphDeployParams,
): Promise<DeployResult> {
  const serializeParams = {
    typegraphPath: params.typegraphPath,
    prefix: params.prefix,
    artifactResolution: true,
    codegen: false,
    prismaMigration: {
      migrationsDir: params.migrationsDir ?? "prisma-migrations",
      migrationActions: Object.entries(params.migrationActions ?? {}),
      defaultMigrationAction: params.defaultMigrationAction ?? {
        apply: true,
        create: false,
        reset: false,
      },
    },
    pretty: false,
  } satisfies SerializeParams;
  const serialized = typegraph.serialize(serializeParams);
  const tgJson = serialized.tgJson;
  const refArtifacts = serialized.ref_artifacts;

  const headers = new Headers();
  headers.append("Content-Type", "application/json");
  const typegate = params.typegate;
  if (typegate.auth) {
    headers.append("Authorization", typegate.auth.asHeaderValue());
  }
  const url = new URL("/typegate", typegate.url);

  // Make sure we have the correct credentials before doing anything
  await pingTypegate(url, headers);

  if (refArtifacts.length > 0) {
    // upload the artifacts
    const artifactUploader = new ArtifactUploader(
      typegate.url,
      refArtifacts,
      typegraph.name,
      typegate.auth,
      headers,
      params.typegraphPath,
    );
    await artifactUploader.uploadArtifacts();
  } else {
    log.debug("no artifacts to upload");
  }

  // deploy the typegraph
  const response = (await execRequest(
    url,
    {
      method: "POST",
      headers,
      body: sdkUtils.gqlDeployQuery({
        tg: tgJson,
        secrets: Object.entries(params.secrets ?? {}),
      }),
    },
    `tgDeploy failed to deploy typegraph ${typegraph.name}`,
    // deno-lint-ignore no-explicit-any
  )) as Record<string, any>;

  if (response.errors) {
    for (const err of response.errors) {
      console.error(err.message);
    }
    throw new Error(`failed to deploy typegraph ${typegraph.name}`);
  }

  const addTypegraph = response.data.addTypegraph;

  /*
  // FIXME: failure field is used by interactive deployment
  // which means errors need to be ignored here but this
  // allows deployment errors in non-interactive scenarios
  if (addTypegraph.failure) {
    console.error(addTypegraph.failure);
    throw new Error(`failed to deploy typegraph ${typegraph.name}`);
  } */

  return {
    serialized: tgJson,
    response: addTypegraph,
  };
}

export async function tgRemove(
  typegraphName: string,
  params: TypegraphRemoveParams,
): Promise<RemoveResult> {
  const { url, auth } = params.typegate;

  const headers = new Headers();
  headers.append("Content-Type", "application/json");
  if (auth) {
    headers.append("Authorization", auth.asHeaderValue());
  }

  const response = (await execRequest(
    new URL("/typegate", url),
    {
      method: "POST",
      headers,
      body: sdkUtils.gqlRemoveQuery([typegraphName.toString()]),
    },
    `tgRemove failed to remove typegraph ${typegraphName}`,
    // deno-lint-ignore no-explicit-any
  )) as Record<string, any> | string;

  return { typegate: response };
}

export async function pingTypegate(url: URL, headers: Headers) {
  await execRequest(
    url,
    {
      method: "POST",
      headers,
      body: sdkUtils.gqlPingQuery(),
    },
    "Failed to access typegate",
  );
}
